### PR TITLE
refactor(builder): remove `unsafe*Builder`s

### PR DESCRIPTION
**Pull request by imranbarbhuiya**
**Please describe the changes this PR makes and why it should be merged:**
This pr removes all the `Unsafe*Builder` in favor of `enableValidators/disableValidators`.
this will resolve 2nd issue of #8015 
**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)
- This PR includes breaking changes (methods removed or renamed, parameters moved or removed)

### CODE REVIEW

File: packages/builders/src/util/validation.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import { setGlobalValidationEnabled } from '@sapphire/shapeshift';
+
+export const enableValidators = () => setGlobalValidationEnabled(true);
```
Comment (vladfrangu) [inline packages/builders/src/util/validation.ts:3]: This is not what I said in the RFC 😓. It should disable *just* builder's assertions, not all validators across runtime that use shapeshift

**Review by vladfrangu — DISMISSED** [review]

File: packages/builders/src/util/validation.ts (hunk 1 above)
Comment (imranbarbhuiya) [reply packages/builders/src/util/validation.ts]: oops. Fixed

**Comment by vladfrangu** [comment]
Again, not what I said. It should not call that setter every time a validation happens. It should call that setter *once* when the disable/enable method is called

**Comment by imranbarbhuiya** [comment]
> Again, not what I said. It should not call that setter every time a validation happens. It should call that setter _once_ when the disable/enable method is called

Copy pasting from discord
> @vladfrangu  u said `enableValidators` should set `setValidationEnabled` to all the predicate (if I've understood correctly then u mean import all the assertion predicates in the validation file and use `setValidationEnabled` on them right?) and not just before parsing but how that'll work? <https://github.com/sapphiredev/shapeshift/blob/a23d15f533a234a59af54a9ed60bec778f069af8/src/validators/BaseValidator.ts#L80> here everytime we run `setValidationEnabled`, it makes a clone so if I add `setValidationEnabled(false)` on a predicate inside of `enableValidators`, it won't change the already present predicate

**Comment by vladfrangu** [comment]
Ooof, rightttt, I made the property protected but the method return a clone... I guess either set the property directly or just keep it as it tbh

**Comment by GravIos3** [comment]
@vladfrangu @iCrawl @kyranet @SpaceEEC Please review

File: packages/builders/src/components/Assertions.ts [hunk 2]
```diff
@@ -1,38 +1,60 @@
...
-export const emojiValidator = s.object({
-	id: s.string,
-	name: s.string,
-	animated: s.boolean,
-}).partial.strict;
+export const emojiValidator = s
+	.object({
+		id: s.string.setValidationEnabled(isValidationEnabled),
+		name: s.string.setValidationEnabled(isValidationEnabled),
+		animated: s.boolean.setValidationEnabled(isValidationEnabled),
```
Comment (vladfrangu) [inline packages/builders/src/components/Assertions.ts:15]: No point in setting these predicate's validation enabled, as the basic typeof checks will still run regardless of the setting

File: packages/builders/src/components/Assertions.ts [hunk 3]
```diff
@@ -1,38 +1,60 @@
...
+	})
+	.partial.strict.setValidationEnabled(isValidationEnabled);
 
-export const disabledValidator = s.boolean;
+export const disabledValidator = s.boolean.setValidationEnabled(isValidationEnabled);
```
Comment (vladfrangu) [inline packages/builders/src/components/Assertions.ts:19]: Ditto

File: packages/builders/src/components/Assertions.ts [hunk 4]
```diff
@@ -1,38 +1,60 @@
...
 
-export const buttonLabelValidator = s.string.lengthGreaterThanOrEqual(1).lengthLessThanOrEqual(80);
+export const buttonLabelValidator = s.string
+	.lengthGreaterThanOrEqual(1)
+	.lengthLessThanOrEqual(80)
+	.setValidationEnabled(isValidationEnabled);
 
-export const buttonStyleValidator = s.nativeEnum(ButtonStyle);
+export const buttonStyleValidator = s.nativeEnum(ButtonStyle).setValidationEnabled(isValidationEnabled);
```
Comment (vladfrangu) [inline packages/builders/src/components/Assertions.ts:26]: Ditto

File: packages/builders/src/components/Assertions.ts [hunk 5]
```diff
@@ -1,38 +1,60 @@
...
+	.lengthLessThanOrEqual(100)
+	.setValidationEnabled(isValidationEnabled);
+export const optionValidator = s
+	.union(
+		s.object({
+			label: labelValueDescriptionValidator,
+			value: labelValueDescriptionValidator,
+			description: labelValueDescriptionValidator.optional,
+			emoji: emojiValidator.optional,
+			default: s.boolean.optional.setValidationEnabled(isValidationEnabled),
```
Comment (vladfrangu) [inline packages/builders/src/components/Assertions.ts:45]: Ditto

File: packages/builders/src/components/selectMenu/SelectMenuOption.ts [hunk 6]
```diff
@@ -8,23 +8,61 @@ import {
...
-		return super.setDescription(labelValueValidator.parse(description));
+export class SelectMenuOptionBuilder {
+	public constructor(public data: Partial<APISelectMenuOption> = {}) {}
+
+	/**
+	 * Sets the label of this option
+	 * @param label The label to show on this option
+	 */
+	public setLabel(label: string) {
+		this.data.label = label;
```
Comment (vladfrangu) [inline packages/builders/src/components/selectMenu/SelectMenuOption.ts:21]: This should validate the label

File: packages/builders/src/components/selectMenu/SelectMenuOption.ts [hunk 7]
```diff
@@ -8,23 +8,61 @@ import {
...
+		return this;
+	}
+
+	/**
+	 * Sets the value of this option
+	 * @param value The value of this option
+	 */
+	public setValue(value: string) {
+		this.data.value = value;
```
Comment (vladfrangu) [inline packages/builders/src/components/selectMenu/SelectMenuOption.ts:30]: This should validate the value

File: packages/builders/src/components/selectMenu/SelectMenuOption.ts [hunk 8]
```diff
@@ -8,23 +8,61 @@ import {
...
+		return this;
+	}
+
+	/**
+	 * Sets the description of this option.
+	 * @param description The description of this option
+	 */
+	public setDescription(description: string) {
+		this.data.description = labelValueValidator.parse(description);
```
Comment (vladfrangu) [inline packages/builders/src/components/selectMenu/SelectMenuOption.ts:39]: Thats not the right validator for this 👀

File: packages/builders/src/components/textInput/Assertions.ts [hunk 9]
```diff
@@ -1,14 +1,24 @@
...
 import { customIdValidator } from '../Assertions';
 
-export const textInputStyleValidator = s.nativeEnum(TextInputStyle);
-export const minLengthValidator = s.number.int.greaterThanOrEqual(0).lessThanOrEqual(4000);
-export const maxLengthValidator = s.number.int.greaterThanOrEqual(1).lessThanOrEqual(4000);
-export const requiredValidator = s.boolean;
-export const valueValidator = s.string.lengthLessThanOrEqual(4000);
-export const placeholderValidator = s.string.lengthLessThanOrEqual(100);
-export const labelValidator = s.string.lengthGreaterThanOrEqual(1).lengthLessThanOrEqual(45);
+export const textInputStyleValidator = s.nativeEnum(TextInputStyle).setValidationEnabled(isValidationEnabled);
```
Comment (vladfrangu) [inline packages/builders/src/components/textInput/Assertions.ts:6]: Ditto

File: packages/builders/src/components/textInput/Assertions.ts [hunk 10]
```diff
@@ -1,14 +1,24 @@
...
+export const minLengthValidator = s.number.int
+	.greaterThanOrEqual(0)
+	.lessThanOrEqual(4000)
+	.setValidationEnabled(isValidationEnabled);
+export const maxLengthValidator = s.number.int
+	.greaterThanOrEqual(1)
+	.lessThanOrEqual(4000)
+	.setValidationEnabled(isValidationEnabled);
+export const requiredValidator = s.boolean.setValidationEnabled(isValidationEnabled);
```
Comment (vladfrangu) [inline packages/builders/src/components/textInput/Assertions.ts:15]: Ditto

File: packages/builders/src/components/textInput/TextInput.ts [hunk 11]
```diff
@@ -7,31 +8,99 @@ import {
...
 	}
 
-	public override setMaxLength(maxLength: number) {
-		return super.setMaxLength(maxLengthValidator.parse(maxLength));
+	/**
+	 * Sets the custom id for this text input
+	 * @param customId The custom id of this text input
+	 */
+	public setCustomId(customId: string) {
+		this.data.custom_id = customId;
```
Comment (vladfrangu) [inline packages/builders/src/components/textInput/TextInput.ts:24]: Missing validation

File: packages/builders/src/components/textInput/TextInput.ts [hunk 12]
```diff
@@ -7,31 +8,99 @@ import {
...
 	}
 
-	public override setRequired(required = true) {
-		return super.setRequired(requiredValidator.parse(required));
+	/**
+	 * Sets the label for this text input
+	 * @param label The label for this text input
+	 */
+	public setLabel(label: string) {
+		this.data.label = label;
```
Comment (vladfrangu) [inline packages/builders/src/components/textInput/TextInput.ts:33]: Missing validation

File: packages/builders/src/components/textInput/TextInput.ts [hunk 13]
```diff
@@ -7,31 +8,99 @@ import {
...
 	}
 
-	public override setValue(value: string) {
-		return super.setValue(valueValidator.parse(value));
+	/**
+	 * Sets the style for this text input
+	 * @param style The style for this text input
+	 */
+	public setStyle(style: TextInputStyle) {
+		this.data.style = style;
```
Comment (vladfrangu) [inline packages/builders/src/components/textInput/TextInput.ts:42]: Missing validation

File: packages/builders/src/interactions/contextMenuCommands/Assertions.ts [hunk 14]
```diff
@@ -1,15 +1,17 @@
...
-
-const typePredicate = s.union(s.literal(ApplicationCommandType.User), s.literal(ApplicationCommandType.Message));
-
-const booleanPredicate = s.boolean;
+	.regex(/^( *[\p{L}\p{N}\p{sc=Devanagari}\p{sc=Thai}_-]+ *)+$/u)
+	.setValidationEnabled(isValidationEnabled);
+const typePredicate = s
+	.union(s.literal(ApplicationCommandType.User), s.literal(ApplicationCommandType.Message))
+	.setValidationEnabled(isValidationEnabled);
+const booleanPredicate = s.boolean.setValidationEnabled(isValidationEnabled);
```
Comment (vladfrangu) [inline packages/builders/src/interactions/contextMenuCommands/Assertions.ts:14]: Ditto

File: packages/builders/src/interactions/contextMenuCommands/Assertions.ts [hunk 15]
```diff
@@ -31,17 +33,19 @@ export function validateRequiredParameters(name: string, type: number) {
 	validateType(type);
 }
 
-const dmPermissionPredicate = s.boolean.nullish;
+const dmPermissionPredicate = s.boolean.nullish.setValidationEnabled(isValidationEnabled);
```
Comment (vladfrangu) [inline packages/builders/src/interactions/contextMenuCommands/Assertions.ts:36]: Ditto

File: packages/builders/src/interactions/contextMenuCommands/Assertions.ts [hunk 16]
```diff
@@ -31,17 +33,19 @@ export function validateRequiredParameters(name: string, type: number) {
...
-	s.number.safeInt.transform((value) => value.toString()),
-	s.string.regex(/^\d+$/),
-).nullish;
+const memberPermissionPredicate = s
+	.union(
+		s.bigint.transform((value) => value.toString()),
+		s.number.safeInt.transform((value) => value.toString()),
+		s.string.regex(/^\d+$/),
+	)
+	.nullish.setValidationEnabled(isValidationEnabled);
```
Comment (vladfrangu) [inline packages/builders/src/interactions/contextMenuCommands/Assertions.ts:48]: Let's not disable validation here, as the constraints also handle transforming the input to valid bitfields (BitInts cannot be JSON stringified, floats are not allowed)

File: packages/builders/src/interactions/slashCommands/Assertions.ts [hunk 17]
```diff
@@ -4,24 +4,29 @@ import { type APIApplicationCommandOptionChoice, Locale, LocalizationMap } from
...
 	namePredicate.parse(name);
 }
 
-const descriptionPredicate = s.string.lengthGreaterThanOrEqual(1).lengthLessThanOrEqual(100);
-const localePredicate = s.nativeEnum(Locale);
+const descriptionPredicate = s.string
+	.lengthGreaterThanOrEqual(1)
+	.lengthLessThanOrEqual(100)
+	.setValidationEnabled(isValidationEnabled);
+const localePredicate = s.nativeEnum(Locale).setValidationEnabled(isValidationEnabled);
```
Comment (vladfrangu) [inline packages/builders/src/interactions/slashCommands/Assertions.ts:23]: Ditto

File: packages/builders/src/interactions/slashCommands/Assertions.ts [hunk 18]
```diff
@@ -45,7 +50,7 @@ export function validateRequiredParameters(
 	validateMaxOptionsLength(options);
 }
 
-const booleanPredicate = s.boolean;
+const booleanPredicate = s.boolean.setValidationEnabled(isValidationEnabled);
```
Comment (vladfrangu) [inline packages/builders/src/interactions/slashCommands/Assertions.ts:53]: Ditto

File: packages/builders/src/messages/embed/Assertions.ts [hunk 19]
```diff
@@ -1,57 +1,84 @@
...
+	.setValidationEnabled(isValidationEnabled);
 
-export const fieldValuePredicate = s.string.lengthGreaterThanOrEqual(1).lengthLessThanOrEqual(1024);
+export const fieldValuePredicate = s.string
+	.lengthGreaterThanOrEqual(1)
+	.lengthLessThanOrEqual(1024)
+	.setValidationEnabled(isValidationEnabled);
 
-export const fieldInlinePredicate = s.boolean.optional;
+export const fieldInlinePredicate = s.boolean.optional.setValidationEnabled(isValidationEnabled);
```
Comment (vladfrangu) [inline packages/builders/src/messages/embed/Assertions.ts:15]: Ditto

File: packages/builders/src/messages/embed/Embed.ts [hunk 20]
```diff
@@ -11,84 +11,227 @@ import {
 	urlPredicate,
 	validateFieldLength,
 } from './Assertions';
-import { EmbedAuthorOptions, EmbedFooterOptions, RGBTuple, UnsafeEmbedBuilder } from './UnsafeEmbed';
 import { normalizeArray, type RestOrArray } from '../../util/normalizeArray';
+export type RGBTuple = [red: number, green: number, blue: number];
```
Comment (vladfrangu) [inline packages/builders/src/messages/embed/Embed.ts:15]:
```suggestion
import { normalizeArray, type RestOrArray } from '../../util/normalizeArray';

export type RGBTuple = [red: number, green: number, blue: number];
```

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: packages/builders/src/components/selectMenu/SelectMenuOption.ts (hunk 6 above)
Comment (imranbarbhuiya) [reply packages/builders/src/components/selectMenu/SelectMenuOption.ts:21]: This is done https://github.com/imranbarbhuiya/discord.js/blob/03ef3873c27eaaa97c1bdba2b0acf35ac0830f08/packages/builders/src/components/selectMenu/SelectMenuOption.ts#L62
here. It was like this before so I kept it as it was.

File: packages/builders/src/components/selectMenu/SelectMenuOption.ts (hunk 7 above)
Comment (imranbarbhuiya) [reply packages/builders/src/components/selectMenu/SelectMenuOption.ts:30]: same as above

File: packages/builders/src/components/selectMenu/SelectMenuOption.ts (hunk 6 above)
Comment (vladfrangu) [reply packages/builders/src/components/selectMenu/SelectMenuOption.ts:21]: Well its also done there to ensure people actually called the set methods that are required. We always want to validate the input the users give in each set method

File: packages/builders/src/components/selectMenu/SelectMenuOption.ts (hunk 8 above)
Comment (imranbarbhuiya) [reply packages/builders/src/components/selectMenu/SelectMenuOption.ts:39]: It was like this before. Ig it's the same for both so I'll just rename it 
https://github.com/discordjs/discord.js/blob/358c3f4a46638fc818e65fcc113d84818bb68c10/packages/builders/src/components/selectMenu/SelectMenuOption.ts#L15

File: packages/builders/src/components/textInput/TextInput.ts (hunk 11 above)
Comment (imranbarbhuiya) [reply packages/builders/src/components/textInput/TextInput.ts:24]: https://github.com/imranbarbhuiya/discord.js/blob/03ef3873c27eaaa97c1bdba2b0acf35ac0830f08/packages/builders/src/components/textInput/TextInput.ts#L92
done here

Comment (vladfrangu) [reply packages/builders/src/components/textInput/TextInput.ts:24]: https://github.com/discordjs/discord.js/pull/8074#discussion_r901412528

File: packages/builders/src/interactions/modals/Modal.ts [hunk 21]
```diff
@@ -1,19 +1,74 @@
...
+	APIActionRowComponent,
+	APIModalActionRowComponent,
+	APIModalInteractionResponseCallbackData,
+} from 'discord-api-types/v10';
 import { titleValidator, validateRequiredParameters } from './Assertions';
-import { UnsafeModalBuilder } from './UnsafeModal';
 import { customIdValidator } from '../../components/Assertions';
+import { ActionRowBuilder, createComponentBuilder, JSONEncodable, ModalActionRowComponentBuilder } from '../../index';
+import { normalizeArray, type RestOrArray } from '../../util/normalizeArray';
+export class ModalBuilder implements JSONEncodable<APIModalInteractionResponseCallbackData> {
```
Comment (kyranet) [inline packages/builders/src/interactions/modals/Modal.ts:10]:
```suggestion
import { normalizeArray, type RestOrArray } from '../../util/normalizeArray';

export class ModalBuilder implements JSONEncodable<APIModalInteractionResponseCallbackData> {
```

File: packages/builders/src/components/textInput/TextInput.ts [hunk 22]
```diff
@@ -1,37 +1,109 @@
...
 	minLengthValidator,
 	placeholderValidator,
 	requiredValidator,
 	valueValidator,
 	validateRequiredParameters,
+	labelValidator,
+	textInputStyleValidator,
 } from './Assertions';
-import { UnsafeTextInputBuilder } from './UnsafeTextInput';
+import { JSONEncodable, isJSONEncodable } from '../../util/jsonEncodable';
```
[the remark this answers is not in the document]
Comment (kyranet) [reply packages/builders/src/components/textInput/TextInput.ts:13]: Also, if possible, try to apply this change everywhere else 😅

**Review by kyranet — CHANGES_REQUESTED** [review]

**Review by kyranet — APPROVED** [review]

File: packages/builders/src/components/button/Button.ts [hunk 23]
```diff
@@ -15,43 +15,87 @@ import {
...
+	 */
+	public setURL(url: string) {
+		(this.data as APIButtonComponentWithURL).url = urlValidator.parse(url);
+		return this;
 	}
 
-	public override setCustomId(customId: string) {
-		return super.setCustomId(customIdValidator.parse(customId));
+	/**
+	 * Sets the custom Id for this button
```
Comment (Jiralite) [inline packages/builders/src/components/button/Button.ts:49]:
```suggestion
	 * Sets the custom id for this button
```

File: packages/builders/src/components/button/Button.ts [hunk 24]
```diff
@@ -15,43 +15,87 @@ import {
...
+	 */
+	public setEmoji(emoji: APIMessageComponentEmoji) {
+		this.data.emoji = emojiValidator.parse(emoji);
+		return this;
 	}
 
-	public override setDisabled(disabled = true) {
-		return super.setDisabled(disabledValidator.parse(disabled));
+	/**
+	 * Sets whether this button is disable or not
```
Comment (Jiralite) [inline packages/builders/src/components/button/Button.ts:69]:
```suggestion
	 * Sets whether this button is disabled or not
```

File: packages/builders/src/components/button/Button.ts [hunk 25]
```diff
@@ -15,43 +15,87 @@ import {
...
+	 *
+	 * @param disabled - Whether or not to disable this button or not
```
Comment (Jiralite) [inline packages/builders/src/components/button/Button.ts:71]:
```suggestion
	 * @param disabled - Whether to disable this button or not
```

File: packages/builders/src/components/selectMenu/SelectMenu.ts [hunk 26]
```diff
@@ -11,61 +10,118 @@ import {
...
+	 */
+	public setMaxValues(maxValues: number) {
+		this.data.max_values = minMaxValidator.parse(maxValues);
+		return this;
 	}
 
-	public override setCustomId(customId: string) {
-		return super.setCustomId(customIdValidator.parse(customId));
+	/**
+	 * Sets the custom Id for this select menu
```
Comment (Jiralite) [inline packages/builders/src/components/selectMenu/SelectMenu.ts:61]:
```suggestion
	 * Sets the custom id for this select menu
```

File: packages/builders/src/components/selectMenu/SelectMenu.ts [hunk 27]
```diff
@@ -11,61 +10,118 @@ import {
...
+	 */
+	public setCustomId(customId: string) {
+		this.data.custom_id = customIdValidator.parse(customId);
+		return this;
 	}
 
-	public override setDisabled(disabled = true) {
-		return super.setDisabled(disabledValidator.parse(disabled));
+	/**
+	 * Sets whether or not this select menu is disabled
```
Comment (Jiralite) [inline packages/builders/src/components/selectMenu/SelectMenu.ts:71]:
```suggestion
	 * Sets whether this select menu is disabled
```

File: packages/builders/src/components/selectMenu/SelectMenu.ts [hunk 28]
```diff
@@ -11,61 +10,118 @@ import {
...
+	 *
+	 * @param disabled - Whether or not this select menu is disabled
```
Comment (Jiralite) [inline packages/builders/src/components/selectMenu/SelectMenu.ts:73]:
```suggestion
	 * @param disabled - Whether this select menu is disabled
```

File: packages/builders/src/components/textInput/TextInput.ts [hunk 29]
```diff
@@ -9,44 +10,108 @@ import {
...
+	 * @param value - The value for this text input
+	 */
+	public setValue(value: string) {
+		this.data.value = valueValidator.parse(value);
+		return this;
 	}
 
-	public override toJSON(): APITextInputComponent {
+	/**
+	 * Sets whether this text input is required or not
```
Comment (Jiralite) [inline packages/builders/src/components/textInput/TextInput.ts:93]:
```suggestion
	 * Sets whether this text input is required
```

File: packages/builders/src/components/textInput/TextInput.ts [hunk 30]
```diff
@@ -9,44 +10,108 @@ import {
...
+	 *
+	 * @param required - Whether this text input is required or not
```
Comment (Jiralite) [inline packages/builders/src/components/textInput/TextInput.ts:95]:
```suggestion
	 * @param required - Whether this text input is required
```

**Review by Jiralite — CHANGES_REQUESTED** [review]

**Comment by iCrawl** [comment]
This needs a rebase.

**Review by SpaceEEC — CHANGES_REQUESTED** [review]
Is it intended that `UnsafeEmbed` is still around? :eyes:

**Comment by imranbarbhuiya** [comment]
> Is it intended that `UnsafeEmbed` is still around? 👀

fixed. It got added while rebasing lol

**Pull request merged**